source/val/ValidationState.h

```diff
@@ -138,36 +144,58 @@ class ValidationState_t {
   /// Keeps track of ID definitions and uses.
   class UseDefTracker {
    public:
-    void AddDef(const spv_id_info_t& def) { defs_[def.id] = def; }
+    explicit UseDefTracker(ValidationState_t* state);
+
+    /// Adds the defintion of an ID
+    void AddDef(const spv_id_info_t& def);
 
-    void AddUse(uint32_t id) { uses_.insert(id); }
+    /// Adds the use of an ID
+    /// @param[in] id      The id of the operand
+    /// @param[in] is_phi  True if the use of the id was part of a phi
+    ///                    instruction
+    void AddUse(uint32_t id, bool is_phi);
 
     /// Finds id's def, if it exists.  If found, returns <true, def>. Otherwise,
     /// returns <false, something>.
-    std::pair<bool, spv_id_info_t> FindDef(uint32_t id) const {
-      if (defs_.count(id) == 0) {
-        return std::make_pair(false, spv_id_info_t{});
-      } else {
-        /// We are in a const function, so we cannot use defs.operator[]().
-        /// Luckily we know the key exists, so defs_.at() won't throw an
-        /// exception.
-        return std::make_pair(true, defs_.at(id));
-      }
-    }
+    std::pair<bool, spv_id_info_t> FindDef(uint32_t id) const;
+
+    /// Propagates the definitions of IDs from the dominators to the blocks they
+    /// dominate for a given function
+    ///
+    /// NOTE: This function requires that the CFG dominators are correctly set
+    /// for each block in the function
+    void UpdateDefs(Function* func);
+
+    /// Returns a map of block IDs that incorrectly used an operand that was not
+    /// first defined by its dominator blocks.
```

Comment:
Return value is unclear.
Is the key the ID that's badly used?  And what's the value for that key?